src/persist/src/indexed/mod.rs

```diff
@@ -1533,4 +1533,40 @@ mod tests {
 
         Ok(())
     }
+
+    #[test]
+    fn clear_indexed_on_invalid_encoding() -> Result<(), Error> {
```

Comment:
There's three relevant cases for us to test
- the version in storage is less than current meta: we should _delete_ the data and then start
- the version in storage matches the current meta: we should _keep_ the data and start
- the version in storage is greater than the current meta: we should _keep_ the data and _not start mz_. (consider the case where we've finished productionizing things. we definitely don't want starting an old version of mz to delete production data)

if I was thinking ahead, I'd have started at version 1 so you could write this test, but I didn't. you'll need to bump to version 1 as part of this PR so you can test the less case

Reply:
So, I didn't do a separate check for if the version is > current because, while I totally agree this is something we'll want when we expose persistent user tables or persistent kafka sources as beta it's less clear to me that we need it right now. The only data currently being persisted is the metrics stuff + experimental user table persistence, neither of which external users care about. On the other hand, users do sometimes cycle between several different versions of Materialize to figure out root cause of some issue (saw this last week with a user going between 0.9.6 - 0.9.10). I'm not sure if adding this sort of check at the moment, which could then require the user to manually clear out `mzdata/persist` is worth it? I guess another implicit assumption I have here is that we won't release anything to beta for a few weeks which is maybe not true.

I don't have a huge preference here, wdyt?

Reply:
I'm sorry, I don't follow. All three cases I outlined above are relevant and necessary from the very moment we introduce this. The actual data being written by persist right now is not relevant to that.

Here's why: say this merges as-is and then goes out in v0.9.12. Then in the future someone is using v0.x.y, where we've started to promise backward compatibility/that we are a source of record. If after using mz for a while, if they accidentally start up on v0.9.12, it will delete all their data. That's totally unacceptable, even if the scenario is contrived.

> I'm not sure if adding this sort of check at the moment, which could then require the user to manually clear out mzdata/persist is worth it?

It only refuses to delete the data if you're starting up on an old version, which we explictly don't support right now anyway. https://materialize.com/docs/versions/#upgrading

Reply:
Ah yeah thats a totally fair point that I didn't consider.

Reply:
> It only refuses to delete the data if you're starting up on an old version, which we explictly don't support right now anyway.

Users end up starting on an older version when they're debugging issues to try to isolate which version led to the issue. That's the only (semi) legit time I've seen people starting up on a old version, and I think the proposed change could make some of that debugging harder, but I absolutely think its worth it given the concern raised of "what about when we say we're a system of record".

Reply:
Oh to be clear, supporting version rollbacks is IMO table stakes for being a production system. I fully expect we'll have to change this policy at some point. However, given that this is the policy as it stands right now, we may as well take advantage of it